Redesign package to be built on top of reusable dataset containers

darsnack opened the pull request:
This works towards #73 by adding low-level dataset containers that are MLUtils.jl compatible out of the box. The end goal is to rewrite the existing datasets to use these containers, but this PR will just add the containers without changing the existing functionality that users expect. In a followup PR, I will rewrite the existing API, provide a deprecation path, and add better documentation (docstrings will still be added here first).

So far, I have ported only the containers from FastAI.jl. I intend to implement the following before merging this PR:

**Low level**
- [x] `FileDataset`: lazy access to a vector of files on disk
- [x] `TableDataset`: any Tables.jl compatible table
- [x] Text: this one I need to decide how to best reuse JuliaText similar to how `TableDataset` wraps Tables.jl
- [x] `HDF5Dataset`: built on HDF5.jl
- [x] `JLD2Dataset`: built on JLD2.jl
- [x] ~~`MATDataset`: built on MAT.jl~~ the data can just be read in as a `Dict` and used with MLUtils.jl without any extra effort (there is no value to having a wrapper here)

**Mid level**
- [x] `CachedDataset`: wraps a lazy dataset like `FileDataset` so that it stays in memory (this is how small datasets like MNIST will be)
- [x] ~~`GroupedDataset`: wraps several named containers together (e.g. for "train" + "val" data)~~ MLUtils already has functionality for this

CarloLucibello commented:
Given https://github.com/JuliaML/MLUtils.jl/pull/56, should we avoid depending on MLUtils and just define `length` and `getindex`?

CarloLucibello commented:
We should drop julia < 1.6

darsnack commented:
The way I wrote `AbstractDataContainer` makes the reverse true as well: anything that supports `getobs`/`numobs` automatically gets `getindex`, `length`, and `iterate`.

If the MLUtils.jl dependency is too expensive, I'd rather move ahead with splitting the interface into its own base package. My view is that MLDatasets should be a first-class consumer of the MLUtils interface. I also think that MLUtils might be a cheaper dep than some of the other IO related packages. Though this is a good question to consider: should there be a "just the data deps" package that splits out only the download/unpack functions?

darsnack commented:
I decided against incorporating text in this PR since it involves a temporal dimension that I haven't yet discussed as part of the MLUtils.jl refactor. I also expect that the best option will involve factoring out text datasets similar to how we discussed for vision.

Would also be useful to get @lorenzoh's eyes on this PR.

CarloLucibello commented on src/containers/cacheddataset.jl:
```diff
@@ -0,0 +1,39 @@
+    source::T
+    cacheidx::Vector{Int}
+    cache::S
+end
```
Should this go in MLUtils.jl?

darsnack replied:
I can make a PR though this is only useful for data that isn't already in memory. I had trouble thinking of cases where that's true but the data isn't a dataset.

CarloLucibello commented on src/containers/cacheddataset.jl:
```diff
@@ -0,0 +1,39 @@
+
+    return isnothing(_i) ? getobs(dataset.source, i) : getobs(dataset.cache, _i)
+end
+MLUtils.numobs(dataset::CachedDataset) = numobs(dataset.source)
```
```suggestion
Base.length(dataset::CachedDataset) = numobs(dataset.source)
```

CarloLucibello commented on src/containers/cacheddataset.jl:
```diff
@@ -0,0 +1,39 @@
+    CachedDataset(source, collect(cacheidx), make_cache(source, cacheidx))
+end
+
+function MLUtils.getobs(dataset::CachedDataset, i::Integer)
```
```suggestion
function Base.getindex(dataset::CachedDataset, i::Integer)
```

CarloLucibello commented on src/containers/cacheddataset.jl:
```diff
@@ -0,0 +1,39 @@
+but your system memory is large enough to store a sizeable chunk of it.
+
+By default the observation indices `1:cachesize` are cached.
+You can manually pass in a `cache` and set of `cacheidx` as well.
```
Not clear what `cache` is supposed to be here.

CarloLucibello commented on src/containers/cacheddataset.jl:
```diff
@@ -0,0 +1,39 @@
+end
+
+function CachedDataset(source, cachesize::Int = numobs(source))
+    cacheidx = 1:cachesize
```
storing the first chachesize entries seems totally arbitrary.

darsnack replied:
Any subset of the indices would be arbitrary. The main reason for including this is to address FFCV from [this issue](https://github.com/JuliaML/MLUtils.jl/issues/13). My lab mate is using FFCV with PyTorch on ImageNet right now. We strongly suspect that the overwhelming majority of the performance gains are from caching a portion of the dataset in memory (my lab mate will request 400GB on our cluster). In this case, the particular indices don't matter so much and how many, so the first N seems as good as any.

CarloLucibello commented on src/containers/filedataset.jl:
```diff
@@ -0,0 +1,57 @@
+FileDataset(dir::Union{AbstractPath, AbstractString}, pattern::AbstractString = "*", depth = 4) =
+    FileDataset(loadfile, dir, pattern, depth)
+
+MLUtils.getobs(dataset::FileDataset, i::Integer) = loadfile(dataset.paths[i])
+MLUtils.getobs(dataset::FileDataset, is::AbstractVector) = map(Base.Fix1(getobs, dataset), is)
+MLUtils.numobs(dataset::FileDataset) = length(dataset.paths)
```
`getindex` and `length`

CarloLucibello commented on src/containers/filedataset.jl:
```diff
@@ -0,0 +1,57 @@
+Recursive glob up to `depth` layers deep within `dir`.
+"""
+function rglob(filepattern = "*", dir = pwd(), depth = 4)
+    patterns = [repeat("*/", i) * filepattern for i in 0:(depth - 1)]
```
does this work on all platforms?

lorenzoh commented on src/containers/filedataset.jl:
```diff
@@ -0,0 +1,57 @@
+function loadfile(file::String)
+    if isimagefile(file)
+        # faster image loading
+        return FileIO.load(file, view = true)
```
I think this special case is no longer necessary with ImageIO.jl now defaulting to JpegTurbo.jl (`view` is ImageMagick specific I think.

lorenzoh commented on src/containers/filedataset.jl:
```diff
@@ -0,0 +1,57 @@
+
+Load a file from disk into the appropriate format.
+"""
+function loadfile(file::String)
```
Not sure if a bunch of nested `if`s is the most elegant solution for this. Since this was meant to be a high-level function for FastAI.jl, maybe we can do without it for now?

darsnack replied:
Should I just go with `FileIO.load` then?

lorenzoh replied:
Yeah, I think that's the best for now. But let's re-export `load` instead of creating a new `loadfile` function

lorenzoh reviewed (commented):
Left some comments, mostly on the uglier bits I put into FastAI.jl.

Beside those things, I think we should standardize on the type we use to represent file paths, i.e. `String` or `AbstractPath`. Even if we want to allow passing both, I think it makes sense to use the  same repr. internally.

In favor of `String`s:

- bit simpler
- the globbing only returns files and I think converting these to `Path`s consumes more memory

In favor of `AbstractPath`s:
- more thoughtful API
- more extensible (e.g. other types of file systems like S3)
- less ambiguous

What do you think?

darsnack commented:
I went with `AbstractString` because it doesn't seem like FilePathsBase.jl is used frequently enough in the ecosystem.

ablaom commented:
@darsnack Do the wrapped `TableDataset` objects themselves implement the Tables.jl interface? That is, is every `TableDataset(some_table)` also a table?

darsnack commented:
Not at this moment, but based on our discussion it's clear that it should! I'll add this tomorrow.

ablaom commented:
If not already, I suggest the wrapper be parameterised by the type of the table being wrapped.  See continuation of same discussion.

darsnack commented:
Yeah for exactly the reasons in the discussion, we already do this. And we have faster paths for when the underlying table is a `DataFrame` or `CSV.File`.

ablaom commented:
I see. That works here, but not ideal if we were to  move this to MLUtils (say)  because we would need DataFrames and CSV as dependencies. Those `getobs` implementations should ideally live in those packages (as does the Tables interface), right?

lorenzoh commented on src/containers/tabledataset.jl:
```diff
@@ -0,0 +1,58 @@
+end
+
+TableDataset(table::T) where {T} = TableDataset{T}(table)
+TableDataset(path::AbstractPath) = TableDataset(DataFrame(CSV.File(path)))
```
```suggestion
TableDataset(path::AbstractString) = TableDataset(DataFrame(CSV.File(path)))
```

lorenzoh replied:
I think this is what is keeping the tests from passing.

CarloLucibello commented:
CI fails due to a missing `using FileIO` I guess. Besides that, this looks good to me

darsnack commented:
Okay this is probably good to go now. `TableDataset` is now a Tables.jl table. We can continue the discussion around tables in MLUtils.jl and update here as needed. But for now, this will allow us to move forward in this repo with the refactor.

CarloLucibello commented:
You can remove  `strict = true` from docs/make.jl if you don't want to add this stuff to the documentation right now.

darsnack commented:
I just added a WIP page so that the docstrings are at least there should someone choose to use the dev branch. I also changed the behavior to only check docstrings for exported functions (since I added `rglob` which we probably don't want in the actual docs).
